dashboard/cron.py

```diff
@@ -48,11 +48,18 @@ def util_function(UDW_course_id, sql_string, mysql_table, table_identifier=None)
 
     # drop duplicates
     df.drop_duplicates(keep='first', inplace=True)
+    
+    logger.debug("Duplicates dropped")
+    logger.debug(df)
 
     logger.debug(" table: " + mysql_table + " insert size: " + str(df.shape[0]))
 
     # write to MySQL
-    df.to_sql(con=engine, name=mysql_table, if_exists='append', index=False)
+    try:
+        df.to_sql(con=engine, name=mysql_table, if_exists='append', index=False)
+    except Exception as e:
+        logger.debug(e)
```

Comment:
use warn level instead of debug, and also log the mysql_table value. 

Not sure whether we can differentiate different Exception types in Python, and handle them differently.

Reply:
Yeah, that makes sense. The cron pod "probably" has debug on but warn is better.

The problem here is there isn't a lot to log. This is all done in the pandas function to_sql. Normally you can differentiate Exception types but the only Exception this looks like it throws is
```
ValueError
When the table already exists and if_exists is ‘fail’ (the default).
```

Everything else is an error from SQL Alchemy and there really is no recovery from this that I could see. It just tries to insert the entire dataframe at once here and if one value fails it all fails. This was failing because we were getting a null value back for one of the primary keys.

We could work around this by inserting one row at a time as suggested like here
https://stackoverflow.com/questions/30337394/pandas-to-sql-fails-on-duplicate-primary-key/35881631

But that would really slow things down. In any case, we cannot have a null as the primary.